cpu-o3: pred: add tage mgsc predictor

jensen-yan opened the pull request:
for ci test

jensen-yan committed:
cpu-o3: pred: add tage mgsc predictor

eastonman commented:
SC and ITTAGE in S3 is acceptable for RTL

jensen-yan committed:
cpu-o3: pred: move sc to s4 stage since tage is in s3 stage.

Change-Id: I437185321799bf9c65e7533f7b8db09bf34a15de

eastonman reviewed (dismissed):
Generally looks good.

jensen-yan committed:
cpu-o3: pred: fix mgsc needMoreHistories

It didn't work before, access to the parent class variables, with a default of 0.

Change-Id: I2a0875dce132f7e4f097e1f332d65c72f66e5be6

jensen-yan committed:
cpu-o3: pred: refactor  MGSC generateSinglePrediction

This commit introduces several utility functions to the BTBMGSC class, including methods to calculate perceptron sums, find weights and thresholds, and determine weight scale differences. These changes aim to reduce code duplication and improve the clarity of the prediction logic.

Change-Id: I3af05333f11d8728d30833b7e0f921bc7f9cbdb1

jensen-yan committed:
cpu: pred: refactor mgsc updateAndAllocateSinglePredictor func

This commit adds several utility functions to the BTBMGSC class, including methods for updating prediction tables, weight tables, and threshold tables.
These changes aim to streamline the update process and improve code clarity by reducing duplication in the update logic.

Change-Id: I79e8abe86cd2d16a570c9759343e78f1c9af3124

jensen-yan committed:
cpu-o3: Update BTBMGSC class with comments and documentation [skip ci]

jensen-yan committed:
cpu-o3: pred: fix imli foled history update

- previously use temp1(temp) is not work, temp is treated as the length of bitset, not the value.
- this fix may have no performance impact since imli is too small for tage.

jensen-yan committed:
cpu-o3: pred: fix sc bias table predition.

previously use main_pred != alt_pred to get biasIndex
now replace alt_pred to alt_pred + tage base pred.
since tage predition is from (main_table, alt_table, base_table)

jensen-yan committed:
cpu-o3: pred: update parameter descriptions in BTBMGSC class

found by github copilot, fix some defines, do not impact performance.

Change-Id: I68337b1c8c4e316a887c502afc812bd4749a4ae0
